h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GamUtils.java

```diff
@@ -0,0 +1,336 @@
+            GamUtils.copy2DArray(genOneGamCol._bInvD, binvD[frameIndex]);
+            numKnotsMat[frameIndex] = genOneGamCol._numKnots;
+            if (nullKnots)  // only copy if knots are not specified
+              System.arraycopy(genOneGamCol._knots, 0, knotsMat[frameIndex], 0, numKnots);
+          }
+        }
+      };
+    }
+    ForkJoinTask.invokeAll(generateGamColumn);
+  }
+
+  public static Frame buildGamFrame(int numGamFrame, Key<Frame>[] gamFramesKey, Frame _train, String response_column) {
+    Vec responseVec=null;
+    if (response_column != null)
+      responseVec = _train.remove(response_column);
+    for (int frameInd = 0; frameInd < numGamFrame; frameInd++) {  // append the augmented columns to _train
+      Frame gamFrame = gamFramesKey[frameInd].get();
+/*      if (centerGAM) {
+        gamColnamesDecenter[frameInd] = new String[gamFrame.names().length];
+        System.arraycopy(gamFrame.names(), 0, gamColnamesDecenter[frameInd], 0, gamFrame.names().length);
+      }*/
+      _train.add(gamFrame.names(), gamFrame.removeAll());
+      Scope.track(gamFrame);
+    }
+    if (response_column!=null)
+      _train.add(response_column, responseVec);
+    return _train;
+  }
+
+  /**
+   * _train now contains original column training columns and added GAM columns
+   * @param numGamFrame
+   * @param gamFramesKey
+   * @param _train
+   * @param response_column
+   * @param trainNCols
+   * @param trainColNames
+   * @return
+   */
+  public static Frame buildGamFrameCenter(int numGamFrame, Key<Frame>[] gamFramesKey, Frame _train,
+                                          String response_column, int trainNCols, String[] trainColNames) {
+    int trainCols = trainNCols-1; // exclude response column
+    Vec[] trainVecs = new Vec[trainCols];
+    int count=0;
+    String[] trainN = new String[trainCols];
+    for (String cname:trainColNames) {  // copy original training columns to gamCenter
+      if (!cname.equals(response_column)) {
+        trainVecs[count] = _train.vec(cname).clone();
+        trainN[count++] = cname;
+      }
+    }
+    Frame gamFrameCenter = new Frame(trainN, trainVecs);
+    for (int frameInd = 0; frameInd < numGamFrame; frameInd++) {  // append the augmented columns to _train
+      Frame gamFrame = gamFramesKey[frameInd].get();
+      gamFrameCenter.add(gamFrame.names(), gamFrame.removeAll());
+      Scope.track(gamFrame);
+    }
+    if (response_column!=null) {
+      gamFrameCenter.add(response_column, _train.vec(response_column));
+    }
+    return gamFrameCenter;
+  }
+
+  public static void addFrameKeys2Keep(List<Key<Vec>> keep, Key<Frame> ... keyNames) {
+    for (Key<Frame> keyName:keyNames) {
+      Frame loadingFrm = DKV.getGet(keyName);
+      if (loadingFrm != null) for (Vec vec : loadingFrm.vecs()) keep.add(vec._key);
+    }
+  }
+
+  public static Frame saveGAMFrames(Frame fr) {
```

Comment:
what is the purpose of this method how is it specific to GAM? or how is it different to new Frame(fr.names(), fr.vecs())?

Reply:
Removed this one since I am storing the whole frame instead of each gam columns.